extensions/guacamole-auth-restrict/src/main/java/org/apache/guacamole/auth/restrict/user/RestrictedUserContext.java

```diff
@@ -0,0 +1,209 @@
+     *
+     * @param userContext
+     *     The UserContext to wrap.
+     * 
+     * @param remoteAddress
+     *     The address the user is logging in from, if known.
+     * 
+     * @param effectiveUserGroups
+     *     The identifiers of the groups this user is associated with.
+     */
+    public RestrictedUserContext(UserContext userContext, String remoteAddress,
+            Set<String> effectiveUserGroups) {
+        super(userContext);
+        this.remoteAddress = remoteAddress;
+        this.effectiveUserGroups = effectiveUserGroups;
+    }
+    
+    @Override
+    public Directory<Connection> getConnectionDirectory() throws GuacamoleException {
+        return new DecoratingDirectory<Connection>(super.getConnectionDirectory()) {
+
+            @Override
+            protected Connection decorate(Connection object) throws GuacamoleException {
+                return new RestrictedConnection(object, remoteAddress);
+            }
+
+            @Override
+            protected Connection undecorate(Connection object) {
+                assert(object instanceof RestrictedConnection);
+                return ((RestrictedConnection) object).getUndecorated();
+            }
+
+        };
+    }
+    
+    @Override
+    public Collection<Form> getConnectionAttributes() {
+        Collection<Form> connectionAttrs = new HashSet<>(super.getConnectionAttributes());
+        connectionAttrs.add(RestrictedConnection.RESTRICT_CONNECTION_FORM);
+        return Collections.unmodifiableCollection(connectionAttrs);
+    }
+    
+    @Override
+    public Directory<ConnectionGroup> getConnectionGroupDirectory() throws GuacamoleException {
+        return new DecoratingDirectory<ConnectionGroup>(super.getConnectionGroupDirectory()) {
+
+            @Override
+            protected ConnectionGroup decorate(ConnectionGroup object) throws GuacamoleException {
+                return new RestrictedConnectionGroup(object, remoteAddress);
+            }
+
+            @Override
+            protected ConnectionGroup undecorate(ConnectionGroup object) {
+                assert(object instanceof RestrictedConnectionGroup);
+                return ((RestrictedConnectionGroup) object).getUndecorated();
+            }
+
+        };
+    }
+    
+    @Override
+    public Collection<Form> getConnectionGroupAttributes() {
+        Collection<Form> connectionGroupAttrs = new HashSet<>(super.getConnectionGroupAttributes());
+        connectionGroupAttrs.add(RestrictedConnectionGroup.RESTRICT_CONNECTIONGROUP_FORM);
+        return Collections.unmodifiableCollection(connectionGroupAttrs);
+    }
+    
+    @Override
+    public Directory<User> getUserDirectory() throws GuacamoleException {
+        return new DecoratingDirectory<User>(super.getUserDirectory()) {
+
+            @Override
+            protected User decorate(User object) throws GuacamoleException {
+                Permissions currentPermissions = self().getEffectivePermissions();
```

Comment:
To avoid potentially requerying the effective permissions of the current user for every other user in the directory, the permission sets involved here should instead be retrieved at entry to `getUserDirectory()`.

Reply:
Yep, makes sense. I've moved it up a couple of lines...

Reply:
I think the calls to retrieve the permission sets (`currentPermissions.getSystemPermissions()` and `currentPermissions.getUserPermissions()`) need to be moved, as well, in case implementations defer retrievals of permissions until the permission type is known.

If possible, it would also be good to avoid calls to `hasPermission()` until specifically needed, such as when invoking `setAttributes()`, as that may also result in a query per user. I think our current JDBC implementation provides a `hasPermission()` that behaves this way, performing one `selectOne()` query per call:

https://github.com/apache/guacamole-client/blob/40620e445cca88c0687e14769b2d7375b3eac703/extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-base/src/main/java/org/apache/guacamole/auth/jdbc/permission/ObjectPermissionSet.java#L100-L104

https://github.com/apache/guacamole-client/blob/40620e445cca88c0687e14769b2d7375b3eac703/extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-base/src/main/java/org/apache/guacamole/auth/jdbc/permission/ModeledObjectPermissionService.java#L173-L187

Alternatively, the entire set of identifiers that are accessible by the user for operations requiring a specific set of permissions can be retrieved in bulk using `getAccessibleObjects()`.

Reply:
Ah, yeah, again, makes sense. I was trying to avoid either having to pass around and store copies of stuff in seemingly unrelated places, or having to implement a bunch of dependency injection just to make that happen, but I'll figure something out.

Reply:
What about pulling the `Set` of accessible identifiers when `getUserDirectory()` is invoked and passing that in instead of a `hasAdmin` flag?

Reply:
Okay, I've done a simpler implementation in the interim that just uses `getAccessibleObjects()` to pull the objects for which the current user has `ADMINISTER` privileges, and moved those calls out of the individual `decorate()` calls so that it only gets called once.

I'll dig into the possibility of using the `Supplier` class - I was having a little bit of trouble with that because it seems as though the `memoize()` method expects to get a function that takes zero arguments, or maybe at most one argument, which would mean needing to do some additional wrapping to pull the `getAccessibleObjects()` method in (would have to write some sort of inline method or something that calls `getAccessibleObjects()` with the `ADMINISTER` type and the identifier set from the user directory. I'll work on that if I have time, but, if the route I went works I'm trying to avoid this holding up the release...

Reply:
Sure - makes sense. I think what you describe sounds good.